internal/controller/apiextensions/usage/reconciler_test.go

```diff
@@ -0,0 +1,776 @@
+						resourceSelectorFn: func(ctx context.Context, u *v1alpha1.Usage) error {
+							return nil
+						},
+					}),
+					WithFinalizer(xpresource.FinalizerFns{AddFinalizerFn: func(_ context.Context, _ xpresource.Object) error {
+						return nil
+					}}),
+				},
+			},
+			want: want{
+				err: errors.Wrap(errBoom, errGetUsing),
+			},
+		},
+		"CannotAddOwnerRefToUsage": {
+			reason: "We should return an error if we cannot add owner reference to the Usage.",
+			args: args{
+				mgr: &fake.Manager{},
+				opts: []ReconcilerOption{
+					WithClientApplicator(xpresource.ClientApplicator{
+						Client: &test.MockClient{
+							MockGet: test.NewMockGetFn(nil, func(obj client.Object) error {
+								if o, ok := obj.(*v1alpha1.Usage); ok {
+									o.Spec.Of.ResourceRef = &v1alpha1.ResourceRef{Name: "used"}
+									o.Spec.By = &v1alpha1.Resource{
+										ResourceRef: &v1alpha1.ResourceRef{Name: "using"},
+									}
+									return nil
+								}
+								if o, ok := obj.(*composed.Unstructured); ok {
+									if o.GetName() == "using" {
+										o.SetAPIVersion("v1")
+										o.SetKind("AnotherKind")
+										o.SetUID("some-uid")
+									}
+									return nil
+								}
+								return errors.New("unexpected object type")
+							}),
+							MockUpdate: test.NewMockUpdateFn(nil, func(obj client.Object) error {
+								if u, ok := obj.(*v1alpha1.Usage); ok {
+									if u.GetOwnerReferences() != nil {
+										return errBoom
+									}
+								}
+								return nil
+							}),
+						},
+					}),
+					WithSelectorResolver(fakeSelectorResolver{
+						resourceSelectorFn: func(ctx context.Context, u *v1alpha1.Usage) error {
+							return nil
+						},
+					}),
+					WithFinalizer(xpresource.FinalizerFns{AddFinalizerFn: func(_ context.Context, _ xpresource.Object) error {
+						return nil
+					}}),
+				},
+			},
+			want: want{
+				err: errors.Wrap(errBoom, errAddOwnerToUsage),
+			},
+		},
+		"SuccessWithUsingResource": {
+			reason: "We should return no error once we have successfully reconciled the usage resource.",
+			args: args{
+				mgr: &fake.Manager{},
+				opts: []ReconcilerOption{
+					WithClientApplicator(xpresource.ClientApplicator{
+						Client: &test.MockClient{
+							MockGet: test.NewMockGetFn(nil, func(obj client.Object) error {
+								if o, ok := obj.(*v1alpha1.Usage); ok {
+									o.Spec.Of.ResourceRef = &v1alpha1.ResourceRef{Name: "used"}
+									o.Spec.By = &v1alpha1.Resource{
+										ResourceRef: &v1alpha1.ResourceRef{Name: "using"},
+									}
+									return nil
+								}
+								if o, ok := obj.(*composed.Unstructured); ok {
+									if o.GetName() == "using" {
+										o.SetAPIVersion("v1")
+										o.SetKind("AnotherKind")
+										o.SetUID("some-uid")
+									}
+									return nil
+								}
+								return errors.New("unexpected object type")
+							}),
+							MockUpdate: test.NewMockUpdateFn(nil, func(obj client.Object) error {
```

Comment:
is the owner reference only thing that gets changed in the successful case? I could see in the reconcile loop that:
* finalizer gets added
* some annexations are added to Usage
* label is added to the resource in use

and perhaps something more? Do we have an unit test that assert all of the above aspects?

Reply:
@pedjak please see https://github.com/crossplane/crossplane/pull/4444#discussion_r1319810642

> Another point that I am just realizing is it is indeed not possible/feasible to do input checks with the current approach we are following. For example, I am trying to add some input checks to an Update call which is called multiple times, and I cannot find a clean way of doing so for different calls 🤷

To use this case as an example, what you're looking for requires checking the state of the Usage resource and resource in use in the `MockUpdate` call which is being called 3 times (instead of 4 because add finalizer call handled by an [xpresource.Finalizer](https://github.com/crossplane/crossplane-runtime/blob/717fe10475b27bdcb8dd4ddfce8364846d9e12e9/pkg/resource/interfaces.go#L102C6-L102C15) which has its own client). The challenge here is, I need to count the calls and validate input based on that. With the current [MockClient](https://github.com/crossplane/crossplane-runtime/blob/717fe10475b27bdcb8dd4ddfce8364846d9e12e9/pkg/test/fake.go#L239) implementation this is not possible/feasible unless I build something dirty. To make it more tangible, my MockUpdate implementation would look like as follows:

```go
MockUpdate: func(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
  switch mockUpdateCalled {
  case 0:
	  o, ok := obj.(*v1alpha1.Usage)
	  if !ok {
		  t.Errorf("expected a Usage but got a %T", obj)
	  }
	  if o.Annotations[detailsAnnotationKey] == "" {
		  t.Errorf("expected finalizer to be set on usage properly")
	  }
  case 1:
	  o, ok := obj.(*composed.Unstructured)
	  if !ok {
		  t.Errorf("expected a composed.Unstructured but got a %T", obj)
	  }
	  if o.GetLabels()[inUseLabelKey] != "true" {
		  t.Errorf("expected %s label to be true", inUseLabelKey)
	  }
  case 2:
	  o, ok := obj.(*v1alpha1.Usage)
	  if !ok {
		  t.Errorf("expected a Usage but got a %T", obj)
	  }
	  if len(o.GetOwnerReferences()) != 1 {
		  t.Errorf("expected owner reference to be set on usage properly")
	  }
	  owner := o.GetOwnerReferences()[0]
	  if owner.APIVersion != "v1" || owner.Kind != "AnotherKind" || owner.UID != "some-uid" {
		  t.Errorf("expected owner reference to be set on usage properly")
	  }
  }
  
  mockUpdateCalled++
  return nil
},
```

I don't like this because it is complicated and fragile. It is adding more complication than the value it is adding, especially now we have e2e tests validating the functionality. 

But I see your point and agree that we are not unit-testing everything we can with the current approach. If we want to go with that route, we should probably switch to something like [gomock](https://pkg.go.dev/github.com/golang/mock/gomock). However, I believe we should discuss this in general, considering this is the approach followed by the whole project as suggested [here](https://github.com/crossplane/crossplane/pull/4444#discussion_r1319811602).